be/src/fs/hdfs/fs_hdfs.cpp

```diff
@@ -418,32 +418,138 @@ class HdfsFileSystem : public FileSystem {
 
     Status iterate_dir2(const std::string& dir, const std::function<bool(DirEntry)>& cb) override;
 
-    Status delete_file(const std::string& path) override { return Status::NotSupported("HdfsFileSystem::delete_file"); }
+    Status delete_file(const std::string& path) override {
+        std::string namenode;
+        RETURN_IF_ERROR(get_namenode_from_path(path, &namenode));
+        std::shared_ptr<HdfsFsClient> hdfs_client;
+        RETURN_IF_ERROR(HdfsFsCache::instance()->get_connection(namenode, hdfs_client, _options));
+
+        if (hdfsDelete(hdfs_client->hdfs_fs, path.c_str(), 0) == 0) {
+            return Status::OK();
+        }
+
+        Status exists_status = _path_exists(hdfs_client->hdfs_fs, path);
+        if (!exists_status.ok()) {
+            return exists_status;
+        }
+
+        ASSIGN_OR_RETURN(bool is_dir, _is_directory(hdfs_client->hdfs_fs, path));
+        if (is_dir) {
+            return Status::InvalidArgument(fmt::format("path {} is a directory, not a file", path));
+        }
+
+        return Status::InternalError(
```

Comment:
internal Error might be too vague for upper level, shall be able to translate known errorno to known Status. Similar to absl::ErrnoToStatus: https://github.com/abseil/abseil-cpp/blob/master/absl/status/status.cc#L404